<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Sibyl • Sibyl</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Sibyl">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Sibyl</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/Sibyl.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Sibyl</h1>
                        <h4 data-toc-skip class="author">Lorenzo
Assentato</h4>
            
            <h4 data-toc-skip class="date">2025-03-04</h4>
      

      <div class="d-none name"><code>Sibyl.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><strong>Sibyl</strong> is a package designed to test different
rarefaction thresholds, when accounting for diverse library size
(sequencing depth) in microbial abundance data.</p>
<p>When performing Principal Component Analysis (PCA), or other types of
ordinations, it is necessary to choose a rarefaction threshold which
does not impact the structure of the data during exploratory
analysis.</p>
<p>The current package builds on an implementation of rarefaction which
corresponds to multiple iterations of random sub-sampling. The result of
a all the repetitions for a single sample are then used to calculate
consensus coordinates for the sample. The functions in the package are
build to accommodate different computational requirements, therefore the
number of times sub-sampling is repeated can be set by the user.</p>
<blockquote>
<p>Keep into account that increasing repeats number requires more
computational power, in particular the main limiting factor is available
memory to store a high number of ordination results.</p>
</blockquote>
<p>There are two key parameters to keep into account while exploring a
dataset with <strong>Sibyl</strong>:</p>
<ul>
<li>
<strong>Rarefaction threshold:</strong> The sequencing depth
threshold at which we decide to sub-sample our data.</li>
<li>
<strong>Repetition number:</strong> this is the number of times we
want to repeat the sub-sampling.</li>
</ul>
<div class="section level3">
<h3 id="data-information-and-source">Data information and source<a class="anchor" aria-label="anchor" href="#data-information-and-source"></a>
</h3>
<p>The data contained in the package and used for illustration comes
from Illumina MiSeq paired-end sequencing of regions V3-V4 of the 16S
genes. The samples used in the study are adult mosquitoes from Burkina
Faso, a more thorough description is in the published paper from Buck et
al. (2016). The other data provided in the package is a collection of
mosquito larvae from Ethiopia from Assentato et al. (2025), which
provides a wider array of sequence depths, from low to high.</p>
</div>
</div>
<div class="section level2">
<h2 id="working-example">Working example<a class="anchor" aria-label="anchor" href="#working-example"></a>
</h2>
<div class="section level3">
<h3 id="sample-completeness-and-sequencing-depth">Sample completeness and sequencing depth<a class="anchor" aria-label="anchor" href="#sample-completeness-and-sequencing-depth"></a>
</h3>
<p>After loading the package the first step is to perform the
calculation of accumulation (rarefaction) curves. This allows a rough
understanding of how the sequencing depth relates to the completeness of
each sample.</p>
<p>The function performs the calculation of rarefaction curves with the
selected step size (a coarser step size require less computational
power).</p>
<p>The <strong>Abundance Coverage Estimator</strong> value (ACE) is
calculated for each sample as a measure of sample richness, and a
generic accumulation curve is fitted to the data, with its asymptote
being the ACE value. This functionality is particularly useful to give a
visual impression of where the curve “break-point” is, or where the
curve is approaching the plateau, fixed as 75% of the ACE value.</p>
<blockquote>
<p>Since in reality these samples never reach their ACE value, which is
always way over their actual sequencing depth (due to it being an
estimation), the “plateau” 75% threshold is to be considered just a
guidance estimate. <strong>It is not an empiric measure</strong>,
therefore should not be treated as such.</p>
</blockquote>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">accumulation_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/accumulation_test.html">accumulation_test</a></span><span class="op">(</span><span class="va">adults</span>, step <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>The resulting list contains:</p>
<ul>
<li>
<strong>accumulation_plot:</strong> Aggregated plot for all samples
with fitted accumulation curves.</li>
<li>
<strong>threshold_density:</strong> 75% thresholds density
plot.</li>
<li>
<strong>individual_plots:</strong> Individual rarefaction plots with
fitted accumulation curves.</li>
</ul>
<p>Let’s take a look at the first 4 individual accumulation plots:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span>  <span class="va">accumulation_result</span><span class="op">$</span><span class="va">individual_plots</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, nrow<span class="op">=</span><span class="fl">2</span>, ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Sibyl_files/figure-html/unnamed-chunk-3-1.png" width="98%" height="500"></p>
<p>The density plot is a visual summary of the range at which most of
the 75% thresholds calculated on the dataset fell.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">accumulation_result</span><span class="op">$</span><span class="va">threshold_density</span></span></code></pre></div>
<p><img src="Sibyl_files/figure-html/unnamed-chunk-4-1.png" width="98%" height="500"></p>
<p>Together, the results produced from this test can be used to choose
an appropriate threshold range to perform the rest of the analysis.</p>
<p>Ideally, while prioritizing sample completeness, we want to make sure
to select a threshold which allows us to keep most of our samples. In
this particular case, to not render the dataset unbalanced.</p>
</div>
<div class="section level3">
<h3 id="running-main-testing-function">Running main testing function<a class="anchor" aria-label="anchor" href="#running-main-testing-function"></a>
</h3>
<p>We proceed by running now the main testing function in the
<strong>Sibyl</strong> package.</p>
<p>Since most of our 75% thresholds, according to the accumulation
analysis, falls in the range between 1000-2000 (and since we are
interested in seeing how low we can push this specific dataset), we
select a range going from 100 to 2000.</p>
<p>Optionally, the repeat number can be a vector of different values. In
that case the testing over different thresholds takes place at different
repeat amounts, and it is possible to directly explore how the amount of
subsampling events affects rarefaction and sample position in the
ordination.</p>
<blockquote>
<p>The choice of a range to test your data on should always be made
considering the compromise between <strong>sample completeness</strong>
(how well described is the sample microbial community according to the
previous analysis), and how many samples are excluded or included when
changing rarefaction threshold.</p>
</blockquote>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">threshold_test_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/test_threshold.html">test_threshold</a></span><span class="op">(</span><span class="va">adults</span>, repeats <span class="op">=</span> <span class="fl">100</span>, t_min <span class="op">=</span> <span class="fl">100</span>, t_max <span class="op">=</span> <span class="fl">2000</span>, t_step <span class="op">=</span> <span class="fl">20</span>, group <span class="op">=</span> <span class="st">"location"</span>, cores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p>This function produces a series of results which can be visualized
immediately or through auxiliary functions.</p>
<p>Those are:</p>
<ol style="list-style-type: decimal">
<li>
<strong>index_plot:</strong> That is probably the most important
result. It shows how the <strong>Calinski-Harabasz index</strong> (CH)
changes when increasing rarefaction threshold. The index is a ratio of
the between-cluster separation (BCSS) to the within-cluster dispersion
(WCSS), normalized by their number of degrees of freedom. A higher index
means that the the between cluster separation is getting bigger and
clusters are less spread out. The index is calculated based on the
consensus points obtained from the sub-sampling attempts. The clusters
are defined as the <strong>“group”</strong> variable used in the
function <strong>“threshold_test_result”</strong>.</li>
</ol>
<p>We are looking for a point in which the CH index is close to a
plateau, which indicates that a higher sequencing effort (a higher
rarefaction threshold) does not affect the ordination.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">threshold_test_result</span><span class="op">$</span><span class="va">index_plot</span></span></code></pre></div>
<p><img src="Sibyl_files/figure-html/unnamed-chunk-6-1.png" width="98%" height="500"></p>
<p>In this other example, we can see what happens when there are samples
in the dataset which do not have enough reads to meet the rarefaction
threshold.</p>
<p>We get a sudden drop in the value of the CH index which indicates
that samples have been removed, and there has been either a decrease of
BCSS or an increase of WCSS (or both). However, this is a good
indication of when samples are getting removed, and if it is important
to keep all samples in the analysis, we might want to operate a choice
that maximize the CH index.</p>
<blockquote>
<p>This has been performed on a different dataset, accessible with:</p>
</blockquote>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">larvae</span><span class="op">)</span></span></code></pre></div>
<p><img src="Sibyl_files/figure-html/unnamed-chunk-9-1.png" width="98%" height="500"></p>
<ol start="2" style="list-style-type: decimal">
<li>
<strong>ordination_plots:</strong> This is a list containing all the
individual ordinations for the selected repetition number across the
whole threshold range to test.</li>
</ol>
<p>Let’s take a look at some of them (lowest threshold, median and
highest):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Isolate the repeat number we need: </span></span>
<span><span class="va">ordination_plots</span> <span class="op">&lt;-</span> <span class="va">threshold_test_result</span><span class="op">$</span><span class="va">ordination_plots</span><span class="op">$</span><span class="va">`repeat_number 100`</span></span></code></pre></div>
<p>It is possible to see how the ordination changes according to the
rarefaction threshold, and how the highest value is not necessarily
making a big difference when compared to the median. The biggest impact
lies in the initial increase of the threshold, but after a critical
point the sample clouds are stable and thus not really affected by the
stochastic sub-sampling procedure.</p>
<p>Thus, this also means that going under a specific rarefaction value
might impact our ordination because of an artificial increase of point
dispersal, which in turn introduces a bias creating an artificial
overlap of two groups which are normally separated.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Print first, middle and last plot:</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ordination_plots</span><span class="op">)</span></span>
<span><span class="co"># Get indices</span></span>
<span><span class="va">indices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="va">n</span> <span class="op">/</span> <span class="fl">2</span><span class="op">)</span>, <span class="va">n</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">selected_plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">indices</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="va">ordination_plots</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">selected_plots</span>, ncol <span class="op">=</span> <span class="fl">3</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Sibyl_files/figure-html/unnamed-chunk-11-1.png" width="98%" height="75%"></p>
<ol start="3" style="list-style-type: decimal">
<li>
<strong>avg_distances:</strong> This is a list of dataframes which
contains the values of average pairwise distances (APD) for each cloud
of points (sub-sampling ordination results) generated from a single
sample.</li>
</ol>
<p>A decreasing trend means that the average pair-wise distance of the
cloud gets smaller.</p>
<p>Measuring APD is especially valuable to give further insight in when
we are reach a point of diminishing returns on the ordination structure.
When increasing the threshold does not increase APD, it means that a
sample position is less affected by repeating the sub-sampling.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">APD_plots</span> <span class="op">&lt;-</span> <span class="va">threshold_test_result</span><span class="op">$</span><span class="va">avg_distances</span><span class="op">$</span><span class="va">repeat_number_100</span></span>
<span><span class="fu"><a href="../reference/avg_pairwise_dist_plot.html">avg_pairwise_dist_plot</a></span><span class="op">(</span><span class="va">APD_plots</span><span class="op">)</span></span></code></pre></div>
<p><img src="Sibyl_files/figure-html/unnamed-chunk-12-1.png" width="98%" height="500"></p>
</div>
<div class="section level3">
<h3 id="consideration-on-the-results">Consideration on the results<a class="anchor" aria-label="anchor" href="#consideration-on-the-results"></a>
</h3>
<p>Overall in this worked example, data suggests that a threshold as low
as 750 does not impact significantly the data structure in the
ordination (according to CH index), and does not affect the single
sample position when repeating sub-sampling (looking at the APD plots).
Despite that, when considering the accumulation curve we can also
consider that 750 reads is most likely not enough to describe accurately
our samples.</p>
<p>However, the package allowed us to explore the effect of different
rarefaction thresholds, and to confirm that lowering a threshold to
include more samples does not necessarily affect our exploratory
analysis, which is still valid for relatively low threshold numbers.</p>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li><p>Buck, M., Nilsson, L., Brunius, C. et al. Bacterial associations
reveal spatial population dynamics in Anopheles gambiae mosquitoes. Sci
Rep 6, 22806 (2016).</p></li>
<li><p>Assentato, L., Nilsson, L., Brunius, C. et al. The type of
environment has a greater impact on the larval microbiota of Anopheles
arabiensis than on the microbiota of their breeding water, FEMS
Microbiology Ecology, Volume 101, Issue 1, (2025).</p></li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lorenzo Assentato, Magdalena Polder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
